File: aoc-2015/src/main/java/com/capital7software/aoc/aoc2015/days/Day23.java
package com.capital7software.aoc.aoc2015.days;

import com.capital7software.aoc.lib.AdventOfCodeSolution;
import com.capital7software.aoc.lib.computer.SimpleComputer;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.time.Instant;
import java.util.List;

/**
 * --- Day 23: Opening the Turing Lock ---<br><br>
 * Little Jane Marie just got her very first computer for Christmas from some unknown benefactor.
 * It comes with instructions and an example program, but the computer itself seems to be malfunctioning.
 * She's curious what the program does, and would like you to help her run it.
 * <p><br>
 * The manual explains that the computer supports two registers and six instructions
 * (truly, it goes on to remind the reader, a state-of-the-art technology).
 * The registers are named a and b, can hold any non-negative integer, and begin with a value of 0.
 * The instructions are as follows:
 * <ul>
 *     <li>
 *         hlf r sets register r to half its current value, then continues with the next instruction.
 *     </li>
 *     <li>
 *         tpl r sets register r to triple its current value, then continues with the next instruction.
 *     </li>
 *     <li>
 *         inc r increments register r, adding 1 to it, then continues with the next instruction.
 *     </li>
 *     <li>
 *         jmp offset is a jump; it continues with the instruction offset away relative to itself.
 *     </li>
 *     <li>
 *         jie r, offset is like jmp, but only jumps if register r is even ("jump if even").
 *     </li>
 *     <li>
 *         jio r, offset is like jmp, but only jumps if register r is 1 ("jump if one", not odd).
 *     </li>
 * </ul>
 * All three jump instructions work with an offset relative to that instruction.<br><br>
 * The offset is always written with a prefix + or - to indicate the direction of the jump
 * (forward or backward, respectively).
 * <p><br>
 * For example, jmp +1 would simply continue with the next instruction, while jmp +0 would
 * continuously jump back to itself forever.
 * <p><br>
 * The program exits when it tries to run an instruction beyond the ones defined.
 * <p><br>
 * For example, this program sets a to 2, because the jio instruction causes it to skip the tpl instruction:
 * <p><br>
 * <code>
 * inc a<br>
 * jio a, +2<br>
 * tpl a<br>
 * inc a<br>
 * </code>
 * <p><br>
 * What is the value in register b when the program in your puzzle input is finished executing?
 * <p><br>
 * Your puzzle answer was 184.
 * <p><br>
 * --- Part Two ---<br><br>
 * The unknown benefactor is very thankful for releasi-- er, helping little Jane Marie with her computer.
 * Definitely not to distract you, what is the value in register b after the program is
 * finished executing if register a starts as 1 instead?
 * <p><br>
 * Your puzzle answer was 231.
 */
public class Day23 implements AdventOfCodeSolution {
    private static final Logger LOGGER = LoggerFactory.getLogger(Day23.class);

    /**
     * Instantiates the solution instance.
     */
    public Day23() {

    }

    @Override
    public String getDefaultInputFilename() {
        return "inputs/input_day_23-01.txt";
    }

    @Override
    public void runPart1(List<String> input) {
        var start = Instant.now();
        var lowest = runProgramAndGetValueInRegister(input, "b");
        var end = Instant.now();
        LOGGER.info("The value in register b is: {}", lowest);
        logTimings(LOGGER, start, end);
    }

    @Override
    public void runPart2(List<String> input) {
        var start = Instant.now();
        var lowest = runProgramWithStartingRegisterAndValueAndGetValueInRegister(
                input,
                "a",
                1L,
                "b"
        );
        var end = Instant.now();
        LOGGER.info("The value in register b is: {}", lowest);
        logTimings(LOGGER, start, end);
    }

    /**
     * Runs the specified program and returns the value in the specified register after the
     * program completes.
     *
     * @param input    The program to run.
     * @param register The register to retrieve the value of after the program runs.
     * @return The value in the specified register after the program completes.
     */
    public long runProgramAndGetValueInRegister(List<String> input, String register) {
        var computer = new SimpleComputer();
        computer.loadProgram(input);
        computer.run();
        return computer.getRegister(register);
    }

    /**
     * Runs the specified program and returns the value in the specified register after the
     * program completes. Prior to execution, the put register is updated with the specified value.
     *
     * @param input       The program to run.
     * @param registerPut The register to set the value of before running the program.
     * @param value       The value to put in the register.
     * @param registerGet The register to retrieve the value of after the program runs.
     * @return The value in the specified register after the program completes.
     */
    public long runProgramWithStartingRegisterAndValueAndGetValueInRegister(
            List<String> input,
            String registerPut,
            Long value,
            String registerGet
    ) {
        var computer = new SimpleComputer();
        computer.loadProgram(input);
        computer.setRegister(registerPut, value);
        computer.run();
        return computer.getRegister(registerGet);
    }
}

File: aoc-2017/src/main/kotlin/com/capital7software/aoc/aoc2017aoc/days/Day23.kt
package com.capital7software.aoc.aoc2017aoc.days

import com.capital7software.aoc.lib.AdventOfCodeSolution
import com.capital7software.aoc.lib.computer.CoProcessor
import com.capital7software.aoc.lib.computer.Instruction
import com.capital7software.aoc.lib.computer.InstructionCounterDebugger
import com.capital7software.aoc.lib.computer.Multiply
import edu.umd.cs.findbugs.annotations.SuppressFBWarnings
import java.time.Instant
import kotlin.reflect.KClass
import org.slf4j.Logger
import org.slf4j.LoggerFactory

/**
 * **--- Day 23: Coprocessor Conflagration ---**
 *
 * You decide to head directly to the CPU and fix the printer from there.
 * As you get close, you find an **experimental coprocessor** doing so much work that the
 * local programs are afraid it will halt and catch fire. This would cause serious issues
 * for the rest of the computer, so you head in and see what you can do.
 *
 * The code it's running seems to be a variant of the kind you saw recently on that tablet.
 * The general functionality seems **very similar**, but some of the instructions are different:
 *
 * - set X Y **sets** register X to the value of Y.
 * - sub X Y **decreases** register X by the value of Y.
 * - mul X Y sets register X to the result of **multiplying** the value contained in
 * register X by the value of Y.
 * - jnz X Y **jumps** with an offset of the value of Y, but only if the value of X
 * is **not zero**. (An offset of 2 skips the next instruction, an offset of -1 jumps
 * to the previous instruction, and so on.)
 *
 * Only the instructions listed above are used. The eight registers here,
 * named a through h, all start at 0.
 *
 * The coprocessor is currently set to some kind of **debug mode**, which allows for testing,
 * but prevents it from doing any meaningful work.
 *
 * If you run the program (your puzzle input), **how many times is the mul instruction invoked?**
 *
 * Your puzzle answer was **5929**.
 *
 * **--- Part Two ---**
 *
 * Now, it's time to fix the problem.
 *
 * The **debug mode switch** is wired directly to register a. You flip the switch, which
 * makes **register a now start at 1** when the program is executed.
 *
 * Immediately, the coprocessor begins to overheat. Whoever wrote this program obviously
 * didn't choose a very efficient implementation. You'll need to **optimize the program** if
 * it has any hope of completing before Santa needs that printer working.
 *
 * The coprocessor's ultimate goal is to determine the final value left in register h
 * once the program completes. Technically, if it had that... it wouldn't even need
 * to run the program.
 *
 * After setting register a to 1, if the program were to run to completion, **what value
 * would be left in register h?**
 *
 * Your puzzle answer was **907**.
 */
class Day23 : AdventOfCodeSolution {
  private companion object {
    private val log: Logger = LoggerFactory.getLogger(Day23::class.java)
  }

  override fun getDefaultInputFilename(): String = "inputs/input_day_23-01.txt"

  override fun runPart1(input: MutableList<String>) {
    val kClass = Multiply::class
    val start = Instant.now()
    val answer = instructionCount(input, kClass)
    val end = Instant.now()

    log.info("$answer is the number of times the $kClass has been executed!")
    logTimings(log, start, end)
  }

  override fun runPart2(input: MutableList<String>) {
    val register = "h"
    val start = Instant.now()
    val answer = registerValue(input, register, 1)
    val end = Instant.now()

    log.info("$answer is the number currently in register $register!")
    logTimings(log, start, end)
  }

  /**
   * Returns the number of times the specified [KClass] has been executed.
   *
   * @param input The raw program instructions.
   * @param kClass The [KClass] of an [Instruction] subclass to get the execution count of.
   * @return The number of times the specified [KClass] has been executed.
   */
  @SuppressFBWarnings
  fun instructionCount(input: List<String>, kClass: KClass<out Instruction>): Long {
    val debugger = InstructionCounterDebugger()
    val instance = CoProcessor.buildCoProcessor(input, debugger)
    debugger.reset()
    instance.reset()
    instance.run()
    val answer = debugger[kClass]
    return answer
  }

  /**
   * Returns the number of times the specified [KClass] has been executed.
   *
   * @param input The raw program instructions.
   * @param register The register to retrieve the value of after program execution.
   * @param a The value to place in register a prior to executing the program.
   * @return The number of times the specified [KClass] has been executed.
   */
  fun registerValue(input: List<String>, register: String, a: Long = 0): Long {
    val instance = CoProcessor.buildCoProcessor(input)
    instance.reset()
    instance["a"] = a
    instance.runOptimized()
    val answer = instance[register]
    return answer
  }
}
